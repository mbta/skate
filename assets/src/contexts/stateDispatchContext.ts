import { createContext } from "react"
import { Dispatch, initialState, State } from "../state"

type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export default StateDispatchContext
