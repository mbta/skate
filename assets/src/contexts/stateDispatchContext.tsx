import React, { createContext, ReactElement } from "react"
import { Dispatch, initialState, State } from "../state"

export type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
export const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export const StateDispatchProvider = ({
  state,
  dispatch,
  children,
}: {
  state: State
  dispatch: Dispatch
  children: ReactElement<HTMLElement>
}) => {
  return (
    <StateDispatchContext.Provider value={[state, dispatch]}>
      {children}
    </StateDispatchContext.Provider>
  )
}
