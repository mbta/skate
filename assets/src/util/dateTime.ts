export const now = (): Date => new Date()

export const dateFromEpochSeconds = (time: number): Date =>
  new Date(time * 1_000)

export const formattedTime = (date: Date): string => {
  const hours24 = date.getHours()
  return `${hours12(hours24)}:${zeroPad(date.getMinutes())}${ampm(hours24)}`
}

export const formattedTimeDiff = (a: Date, b: Date): string => {
  const diffMs: number = a.valueOf() - b.valueOf()
  const diffHours = Math.floor(diffMs / 3_600_000)
  const diffMinutes = Math.floor((diffMs % 3_600_000) / 60_000)

  const diffMinutesStr = `${diffMinutes}m`

  return diffHours >= 1 ? `${diffHours}h ${diffMinutesStr}` : diffMinutesStr
}

/** Takes a time of day in seconds since midnight
 */
export const formattedScheduledTime = (time: number): string => {
  const minutes = Math.floor(time / 60)
  const hours25 = Math.floor(minutes / 60)
  const minutes60 = minutes - hours25 * 60
  return `${hours12(hours25)}:${zeroPad(minutes60)}${ampm(hours25)}`
}

export const hours12 = (hours25: number): number => hours25 % 12 || 12

export const ampm = (hours24: number): string =>
  hours24 >= 12 && hours24 < 24 ? "pm" : "am"

const zeroPad = (time: number): string => (time < 10 ? `0${time}` : `${time}`)
