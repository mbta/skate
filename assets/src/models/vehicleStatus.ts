import featureIsEnabled from "../laboratoryFeatures"
import { Vehicle, VehicleOrGhost } from "../realtime.d"
import { isGhost, isVehicle } from "./vehicle"
import { VehicleAdherenceColorsSetting } from "../userSettings"

/** Where a vehicle is relative to its schedule.
 * Swiftly assigns this to all vehicles.
 */
export type OnTimeStatus = "on-time" | "early" | "late"

export enum HeadwaySpacing {
  VeryBunched = "very-bunched",
  Bunched = "bunched",
  Ok = "ok",
  Gapped = "gapped",
  VeryGapped = "very-gapped",
}

/** How the vehicle triangle should be drawn,
 * taking into account the vehicle's state and the settings
 */
export type DrawnStatus = OnTimeStatus | "off-course" | "ghost" | "plain"

export const onTimeStatus = (scheduleAdherenceSecs: number): OnTimeStatus => {
  const oneMinuteInSeconds = 60
  const sixMinutesInSeconds = 360

  if (scheduleAdherenceSecs < -oneMinuteInSeconds) {
    return "early"
  } else if (scheduleAdherenceSecs > sixMinutesInSeconds) {
    return "late"
  } else {
    return "on-time"
  }
}

export const drawnStatus = (vehicleOrGhost: VehicleOrGhost): DrawnStatus => {
  if (isGhost(vehicleOrGhost)) {
    return "ghost"
  }
  const vehicle = vehicleOrGhost

  if (isVehicle(vehicle) && vehicle.isShuttle) {
    return "plain"
  }

  if (vehicle.isOffCourse) {
    return "off-course"
  }

  if (
    featureIsEnabled("headway_ladder_colors") &&
    vehicle.headwaySpacing !== null
  ) {
    // Headway lines give the status instead of the vehicles
    return "plain"
  }

  return onTimeStatus(vehicle.scheduleAdherenceSecs)
}

export const humanReadableScheduleAdherence = (vehicle: Vehicle): string =>
  vehicle.isOffCourse
    ? "Invalid"
    : humanReadableOnTimeStatus(onTimeStatus(vehicle.scheduleAdherenceSecs))

const humanReadableOnTimeStatus = (status: OnTimeStatus): string => {
  switch (status) {
    case "early":
      return "early"

    case "on-time":
      return "on time"

    case "late":
      return "late"
  }
}

export const humanReadableHeadwaySpacing = (
  spacing: HeadwaySpacing | null
): string => {
  switch (spacing) {
    case null:
      return "good"

    case HeadwaySpacing.VeryGapped:
      return "very gapped"

    case HeadwaySpacing.Gapped:
      return "gapped"

    case HeadwaySpacing.Ok:
      return "good"

    case HeadwaySpacing.Bunched:
      return "bunched"

    case HeadwaySpacing.VeryBunched:
      return "very bunched"
  }
}

export const headwaySpacingToString = (spacing: HeadwaySpacing): string => {
  switch (spacing) {
    case HeadwaySpacing.VeryGapped:
      return "very-gapped"

    case HeadwaySpacing.Gapped:
      return "gapped"

    case HeadwaySpacing.Ok:
      return "ok"

    case HeadwaySpacing.Bunched:
      return "bunched"

    case HeadwaySpacing.VeryBunched:
      return "very-bunched"
  }
}

export const statusClasses = (
  status: DrawnStatus,
  vehicleAdherenceColors: VehicleAdherenceColorsSetting
): string[] => {
  if (status === "plain") {
    return [""]
  } else {
    switch (vehicleAdherenceColors) {
      case VehicleAdherenceColorsSetting.EarlyRed:
        return [status, "early-red"]
      case VehicleAdherenceColorsSetting.EarlyBlue:
        return [status, "early-blue"]
    }
  }
}
