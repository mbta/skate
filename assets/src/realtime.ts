import {
  BlockId,
  DirectionId,
  RouteId,
  RoutePatternId,
  StopId,
  TimepointId,
  TripId,
  ViaVariant,
} from "./schedule"

import { Crowding } from "./models/crowding"
import { DetourId } from "./models/detoursList"

export interface BlockWaiver {
  startTime: Date
  endTime: Date
  causeId: number
  causeDescription: string
  remark: string | null
}

export interface DataDiscrepancy {
  attribute: string
  sources: DataDiscrepancySource[]
}

export interface DataDiscrepancySource {
  id: SourceId
  value: string | null
}

export type EndOfTripType = "another_trip" | "swing_off" | "pull_back"

export interface Ghost {
  id: VehicleId
  directionId: DirectionId
  routeId: RouteId
  routePatternId: RoutePatternId | null
  tripId: TripId
  headsign: string
  blockId: BlockId
  runId: RunId | null
  viaVariant: ViaVariant | null
  incomingTripDirectionId: DirectionId | null
  layoverDepartureTime: number | null
  scheduledTimepointStatus: VehicleTimepointStatus | null
  scheduledLogonTime: number | null
  routeStatus: RouteStatus
  blockWaivers: BlockWaiver[]
  currentPieceStartPlace: string | null
  currentPieceFirstRoute: RouteId | null
}

export type NotificationId = string

// A Notification's `content` is populated with an Elixir struct and the
// `__struct__` field is serialized so that the type can be discerned on the
// frontend.
export enum NotificationType {
  BridgeMovement = "Elixir.Notifications.Db.BridgeMovement",
  BlockWaiver = "Elixir.Notifications.Db.BlockWaiver",
  Detour = "Elixir.Notifications.Db.Detour",
}

export interface BridgeLoweredNotification {
  $type: NotificationType.BridgeMovement
  status: "lowered"
}

export interface BridgeRaisedNotification {
  $type: NotificationType.BridgeMovement
  status: "raised"
  loweringTime: Date
}

export type BridgeNotification =
  | BridgeLoweredNotification
  | BridgeRaisedNotification

export type BlockWaiverNotification = {
  $type: NotificationType.BlockWaiver
  createdAt: Date
  reason: BlockWaiverReason
  routeIds: RouteId[]
  runIds: RunId[]
  tripIds: TripId[]
  operatorId: string | null
  operatorName: string | null
  routeIdAtCreation: string | null
  startTime: Date
  endTime: Date | null
}

export enum DetourNotificationStatus {
  Activated = "activated",
  Deactivated = "deactivated",
}

export type DetourNotification = {
  $type: NotificationType.Detour
  status: DetourNotificationStatus
  detourId: DetourId
  headsign: string
  route: string
  direction: string
  origin: string
}

export type NotificationContentTypes =
  | BridgeNotification
  | BlockWaiverNotification
  | DetourNotification

export interface Notification<
  TNotification extends NotificationContentTypes = NotificationContentTypes
> {
  id: NotificationId
  createdAt: Date
  state: NotificationState
  content: TNotification
}

export function isBlockWaiverNotification(
  notification: Notification
): notification is Notification<BlockWaiverNotification> {
  return notification.content.$type === NotificationType.BlockWaiver
}
export type BlockWaiverReason =
  | "other"
  | "accident"
  | "adjusted"
  | "disabled"
  | "diverted"
  | "manpower"
  | "operator_error"
  | "traffic"

export type NotificationState = "unread" | "read" | "deleted"

export type SourceId = string

export type RouteStatus = "on_route" | "laying_over" | "pulling_out"

export type RunId = string

export interface ByRunId<T> {
  [runId: string]: T
}

export type TrainVehicleId = string

export interface TrainVehicle {
  id: TrainVehicleId
  latitude: number
  longitude: number
  bearing: number
}

export interface Vehicle {
  id: VehicleId
  label: string | null
  runId: RunId | null
  timestamp: number
  latitude: number
  longitude: number
  directionId: DirectionId | null
  routeId: RouteId | null
  routePatternId: RoutePatternId | null
  tripId: TripId | null
  headsign: string | null
  viaVariant: ViaVariant | null
  operatorId: string | null
  operatorFirstName: string | null
  operatorLastName: string | null
  operatorLogonTime: Date | null
  overloadOffset?: number
  bearing?: number
  blockId: BlockId | null
  previousVehicleId: string | null
  scheduleAdherenceSecs: number | null
  incomingTripDirectionId: DirectionId | null
  isShuttle: boolean
  isOverload: boolean
  isOffCourse: boolean
  isRevenue: boolean | null
  layoverDepartureTime: number | null
  pullbackPlaceName: string | null
  dataDiscrepancies: DataDiscrepancy[]
  stopStatus: VehicleStopStatus
  timepointStatus: VehicleTimepointStatus | null
  scheduledLocation: VehicleScheduledLocation | null
  routeStatus: RouteStatus
  endOfTripType: EndOfTripType
  blockWaivers: BlockWaiver[]
  crowding: Crowding | null
}

export interface VehicleInScheduledService extends Vehicle {
  directionId: DirectionId
  routeId: RouteId
  blockId: BlockId
}

export type VehicleId = string

export interface VehicleStopStatus {
  stopId: StopId | null
  stopName: string | null
}

export interface VehicleScheduledLocation {
  routeId: RouteId
  routePatternId: RoutePatternId | null
  directionId: DirectionId
  tripId: TripId
  runId: RunId
  timeSinceTripStartTime: number
  headsign: string | null
  viaVariant: ViaVariant | null
  timepointStatus: VehicleTimepointStatus
}

export interface VehicleTimepointStatus {
  timepointId: TimepointId
  fractionUntilTimepoint: number
}
