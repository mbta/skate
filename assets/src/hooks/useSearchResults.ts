import { Channel, Socket } from "phoenix"
import { useEffect, useState } from "react"
import { Search } from "../models/search"
import {
  VehicleOrGhostData,
  vehicleOrGhostFromData,
} from "../models/vehicleData"
import { VehicleOrGhost } from "../realtime"

interface SearchResultsPayload {
  data: VehicleOrGhostData[]
}

const subscribe = (
  socket: Socket,
  search: Search,
  setVehicles: (vehicles: VehicleOrGhost[]) => void
): Channel => {
  const handleSearchResults = (payload: SearchResultsPayload): void => {
    setVehicles(
      payload.data.map((data: VehicleOrGhostData) =>
        vehicleOrGhostFromData(data)
      )
    )
  }

  const channel = socket.channel(
    `vehicles:search:${search.property}:${search.text}`
  )
  channel.on("search", handleSearchResults)

  channel
    .join()
    .receive("ok", handleSearchResults)
    .receive("error", ({ reason }) =>
      // tslint:disable-next-line: no-console
      console.error("search channel join failed", reason)
    )
    .receive("timeout", () => {
      window.location.reload(true)
    })

  return channel
}

const useSearchResults = (
  socket: Socket | undefined,
  search: Search
): VehicleOrGhost[] | null | undefined => {
  const [vehicles, setVehicles] = useState<VehicleOrGhost[] | null | undefined>(
    undefined
  )
  const [channel, setChannel] = useState<Channel | undefined>(undefined)

  const leaveChannel = () => {
    if (channel !== undefined) {
      channel.leave()
      setChannel(undefined)
    }
  }

  useEffect(() => {
    if (!search.isActive) {
      leaveChannel()
      setVehicles(undefined)
    }

    if (socket && search.isActive) {
      setVehicles(null)
      setChannel(subscribe(socket, search, setVehicles))
    }

    return leaveChannel
  }, [socket, search])

  return vehicles
}

export default useSearchResults
