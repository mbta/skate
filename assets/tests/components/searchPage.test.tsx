import { mount } from "enzyme"
import React from "react"
import renderer from "react-test-renderer"
import SearchPage from "../../src/components/searchPage"
import { StateDispatchProvider } from "../../src/contexts/stateDispatchContext"
import useSearchResults from "../../src/hooks/useSearchResults"
import { HeadwaySpacing } from "../../src/models/vehicleStatus"
import { Ghost, Vehicle, VehicleOrGhost } from "../../src/realtime"
import { initialState, State } from "../../src/state"
import * as dateTime from "../../src/util/dateTime"

jest
  .spyOn(dateTime, "now")
  .mockImplementation(() => new Date("2018-08-15T17:41:21.000Z"))

jest.spyOn(Date, "now").mockImplementation(() => 234000)

const vehicle: Vehicle = {
  id: "v1",
  label: "v1-label",
  runId: "run-1",
  timestamp: 123,
  latitude: 0,
  longitude: 0,
  directionId: 0,
  routeId: "39",
  tripId: "t1",
  headsign: "Forest Hills",
  viaVariant: "X",
  operatorId: "op1",
  operatorName: "SMITH",
  operatorLogonTime: new Date("2018-08-15T13:38:21.000Z"),
  bearing: 33,
  blockId: "block-1",
  headwaySecs: 859.1,
  headwaySpacing: HeadwaySpacing.Ok,
  previousVehicleId: "v2",
  scheduleAdherenceSecs: 0,
  scheduledHeadwaySecs: 120,
  isShuttle: false,
  isOffCourse: false,
  layoverDepartureTime: null,
  blockIsActive: false,
  dataDiscrepancies: [
    {
      attribute: "trip_id",
      sources: [
        {
          id: "swiftly",
          value: "swiftly-trip-id",
        },
        {
          id: "busloc",
          value: "busloc-trip-id",
        },
      ],
    },
  ],
  stopStatus: {
    stopId: "s1",
    stopName: "Stop Name",
  },
  timepointStatus: {
    timepointId: "tp1",
    fractionUntilTimepoint: 0.5,
  },
  scheduledLocation: null,
  routeStatus: "on_route",
  endOfTripType: "another_trip",
  blockWaivers: [],
}
const ghost: Ghost = {
  id: "ghost-trip",
  directionId: 0,
  routeId: "39",
  tripId: "trip",
  headsign: "headsign",
  blockId: "block",
  runId: "123-0123",
  viaVariant: "X",
  layoverDepartureTime: null,
  scheduledTimepointStatus: {
    timepointId: "t0",
    fractionUntilTimepoint: 0.0,
  },
  routeStatus: "on_route",
  blockWaivers: [],
}
jest.mock("../../src/hooks/useSearchResults", () => ({
  __esModule: true,
  default: jest.fn(),
}))

describe("SearchPage", () => {
  test("renders the empty state", () => {
    ;(useSearchResults as jest.Mock).mockImplementationOnce(() => null)
    const tree = renderer
      .create(
        <StateDispatchProvider state={initialState} dispatch={jest.fn()}>
          <SearchPage />
        </StateDispatchProvider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders vehicle data", () => {
    const searchResults: VehicleOrGhost[] = [vehicle, ghost]
    ;(useSearchResults as jest.Mock).mockImplementation(() => searchResults)
    const tree = renderer
      .create(
        <StateDispatchProvider state={initialState} dispatch={jest.fn()}>
          <SearchPage />
        </StateDispatchProvider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a selected vehicle", () => {
    const selectedVehicleState: State = {
      ...initialState,
      selectedVehicleId: "v1",
    }

    const tree = renderer
      .create(
        <StateDispatchProvider
          state={selectedVehicleState}
          dispatch={jest.fn()}
        >
          <SearchPage />
        </StateDispatchProvider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("on mobile, shows the results list initially", () => {
    const wrapper = mount(<SearchPage />)

    expect(wrapper.exists(".m-search-page--show-list")).toBeTruthy()
  })

  test("on mobile, allows you to toggle to the map view and back again", () => {
    const wrapper = mount(<SearchPage />)

    wrapper
      .find(".m-search-page__toggle-mobile-display-button")
      .simulate("click")

    expect(wrapper.exists(".m-search-page--show-map")).toBeTruthy()

    wrapper
      .find(".m-search-page__toggle-mobile-display-button")
      .simulate("click")

    expect(wrapper.exists(".m-search-page--show-list")).toBeTruthy()
  })
})
