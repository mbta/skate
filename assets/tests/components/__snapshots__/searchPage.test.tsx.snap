// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage renders a selected vehicle 1`] = `
<div
  className="c-page m-search-page m-search-page--show-list"
>
  <div
    className="m-search-page__input-and-results"
  >
    <div
      className="m-search-page__input"
    >
      <form
        className="m-search-form"
        onSubmit={[Function]}
      >
        <div
          className="m-search-form__row"
        >
          <div
            className="m-search-form__text"
          >
            <input
              autoFocus={true}
              className="m-search-form__input"
              onChange={[Function]}
              placeholder="Search"
              type="text"
              value=""
            />
            <button
              className="m-search-form__clear"
              onClick={[Function]}
              type="reset"
            >
              <span
                className=""
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "SVG",
                  }
                }
              />
            </button>
          </div>
          <button
            className="m-search-form__submit"
            disabled={true}
            onClick={[Function]}
            type="submit"
          >
            <span
              className=""
              dangerouslySetInnerHTML={
                Object {
                  "__html": "SVG",
                }
              }
            />
          </button>
        </div>
        <div
          className="m-search-form__row"
        >
          <ul
            className="m-search-form__property-buttons"
          >
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={true}
                className="m-search-form__property-input"
                id="property-all"
                name="property"
                onChange={[Function]}
                type="radio"
                value="all"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-all"
              >
                all
              </label>
            </li>
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={false}
                className="m-search-form__property-input"
                id="property-run"
                name="property"
                onChange={[Function]}
                type="radio"
                value="run"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-run"
              >
                run
              </label>
            </li>
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={false}
                className="m-search-form__property-input"
                id="property-vehicle"
                name="property"
                onChange={[Function]}
                type="radio"
                value="vehicle"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-vehicle"
              >
                vehicle
              </label>
            </li>
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={false}
                className="m-search-form__property-input"
                id="property-operator"
                name="property"
                onChange={[Function]}
                type="radio"
                value="operator"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-operator"
              >
                operator
              </label>
            </li>
          </ul>
        </div>
      </form>
      <button
        className="m-search-page__toggle-mobile-display-button"
        onClick={[Function]}
      >
        Show 
        map
         instead
      </button>
    </div>
    <div
      className="m-search-display"
    >
      <div
        className="m-recent-searches"
      >
        <div
          className="m-recent-searches__heading"
        >
          Recent Searches
        </div>
      </div>
    </div>
  </div>
  <div
    className="m-search-page__map"
  >
    <div
      className="m-vehicle-map-state m-vehicle-map-state--auto-centering"
    />
    <div
      className="m-vehicle-map"
      id="id-vehicle-map"
    />
  </div>
  <div
    className="m-properties-panel"
    id="m-properties-panel"
  >
    <div
      className="m-vehicle-properties-panel"
    >
      <div
        className="m-properties-panel__header-wrapper"
      >
        <div
          className="m-properties-panel__header"
        >
          <div
            className="m-properties-panel__label"
          >
            <svg
              style={
                Object {
                  "height": 64,
                  "width": 64,
                }
              }
              viewBox="-32 -20 64 64"
            >
              <g
                className="m-vehicle-icon m-vehicle-icon--large on-time early-red"
              >
                <rect
                  className="m-vehicle-icon__label-background"
                  height={24}
                  rx={12}
                  ry={12}
                  width={64}
                  x={-32}
                  y={19}
                />
                <text
                  className="m-vehicle-icon__label m-vehicle-icon__label--normal"
                  dominantBaseline="central"
                  textAnchor="middle"
                  x="0"
                  y={31}
                >
                  1
                </text>
                <path
                  className="m-vehicle-icon__triangle"
                  d="m27.34 9.46 16.84 24.54a4.06 4.06 0 0 1 -1 5.64 4.11 4.11 0 0 1 -2.3.71h-33.72a4.06 4.06 0 0 1 -4.06-4.11 4 4 0 0 1 .72-2.24l16.84-24.54a4.05 4.05 0 0 1 5.64-1.05 4 4 0 0 1 1.04 1.05z"
                  transform="scale(1) rotate(0) translate(-24,-22)"
                />
                <text
                  className="m-vehicle-icon__variant"
                  dominantBaseline="alphabetic"
                  textAnchor="middle"
                  x={0}
                  y={14}
                >
                  X
                </text>
              </g>
            </svg>
          </div>
          <div
            className="m-properties-panel__variant"
          >
            <div
              className="m-properties-panel__inbound-outbound"
            >
              
            </div>
            <div
              className="m-route-variant-name"
            >
              <span
                className="m-route-variant-name__route-id"
              >
                39_X
              </span>
               Forest Hills
            </div>
            <div
              className="m-properties-panel__schedule-adherence on-time early-red"
            >
              <div
                className="m-properties-panel__schedule-adherence-status-icon"
              >
                <svg
                  height="10"
                  width="10"
                >
                  <circle
                    cx="5"
                    cy="5"
                    r="5"
                  />
                </svg>
              </div>
              <div
                className="m-properties-panel__schedule-adherence-status-string"
              >
                on time
              </div>
              <div
                className="m-properties-panel__schedule-adherence-label"
              >
                (0 min early)
              </div>
            </div>
          </div>
          <div
            className="m-properties-panel__close-ping"
          >
            <button
              className="m-close-button"
              onClick={[Function]}
            >
              <span
                className=""
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "SVG",
                  }
                }
              />
            </button>
            <div
              className="m-properties-panel__last-gps-ping"
            >
              111s ago
            </div>
          </div>
        </div>
        <ul
          className="m-tabs__tab-list"
        >
          <li
            className="m-tabs__tab m-tabs__tab--selected"
            onClick={[Function]}
          >
            <svg
              className="m-tabs__tab-status-icon"
              height="14"
              width="14"
            >
              <circle
                className="m-tabs__tab-status-icon-outer-circle"
                cx="7"
                cy="7"
                r="6"
              />
              <circle
                className="m-tabs__tab-status-icon-inner-circle"
                cx="7"
                cy="7"
                r="4"
              />
            </svg>
            Status
          </li>
          <li
            className="m-tabs__tab"
            onClick={[Function]}
          >
            <svg
              className="m-tabs__tab-status-icon"
              height="14"
              width="14"
            >
              <circle
                className="m-tabs__tab-status-icon-outer-circle"
                cx="7"
                cy="7"
                r="6"
              />
              <circle
                className="m-tabs__tab-status-icon-inner-circle"
                cx="7"
                cy="7"
                r="4"
              />
            </svg>
            Run
          </li>
          <li
            className="m-tabs__tab"
            onClick={[Function]}
          >
            <svg
              className="m-tabs__tab-status-icon"
              height="14"
              width="14"
            >
              <circle
                className="m-tabs__tab-status-icon-outer-circle"
                cx="7"
                cy="7"
                r="6"
              />
              <circle
                className="m-tabs__tab-status-icon-inner-circle"
                cx="7"
                cy="7"
                r="4"
              />
            </svg>
            Block
          </li>
        </ul>
      </div>
      <div
        className="m-tabs__tab-panel"
      >
        <div
          className="m-properties-list"
        >
          <table
            className="m-properties-list__table"
          >
            <tbody>
              <tr
                className="m-properties-list__property "
              >
                <td
                  className="m-properties-list__property-label"
                >
                  Run
                </td>
                <td
                  className="m-properties-list__property-value"
                >
                  run-1
                </td>
              </tr>
              <tr
                className="m-properties-list__property "
              >
                <td
                  className="m-properties-list__property-label"
                >
                  Vehicle
                </td>
                <td
                  className="m-properties-list__property-value"
                >
                  v1-label
                </td>
              </tr>
              <tr
                className="m-properties-list__property "
              >
                <td
                  className="m-properties-list__property-label"
                >
                  Operator
                </td>
                <td
                  className="m-properties-list__property-value"
                >
                  SMITH #op1
                </td>
              </tr>
              <tr
                className="m-properties-list__property m-properties-list__property--last-login"
              >
                <td
                  className="m-properties-list__property-label"
                >
                  Last Login
                </td>
                <td
                  className="m-properties-list__property-value"
                >
                  4 hr 3 min; 1:38 PM
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          className="m-vehicle-properties-panel__location"
        >
          <div
            className="m-vehicle-properties-panel__latlng"
          >
            <div
              className="m-vehicle-properties-panel__label"
            >
              Current Location
            </div>
            <a
              className="m-vehicle-properties-panel__link"
              href="https://www.google.com/maps/dir/?api=1&destination=0,0&travelmode=driving"
              target="_blank"
            >
              Directions
            </a>
          </div>
          <div
            className="m-vehicle-properties-panel__next-stop"
          >
            <div
              className="m-vehicle-properties-panel__label"
            >
              Next Stop
            </div>
            <div
              className="m-vehicle-properties-panel__value"
            >
              Stop Name
            </div>
          </div>
          <div
            className="m-vehicle-properties-panel__map"
          >
            <div
              className="m-vehicle-map-state m-vehicle-map-state--auto-centering"
            />
            <div
              className="m-vehicle-map"
              id="id-vehicle-map"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="m-properties-panel__modal-overlay"
    onClick={[Function]}
  />
</div>
`;

exports[`SearchPage renders the empty state 1`] = `
<div
  className="c-page m-search-page m-search-page--show-list"
>
  <div
    className="m-search-page__input-and-results"
  >
    <div
      className="m-search-page__input"
    >
      <form
        className="m-search-form"
        onSubmit={[Function]}
      >
        <div
          className="m-search-form__row"
        >
          <div
            className="m-search-form__text"
          >
            <input
              autoFocus={true}
              className="m-search-form__input"
              onChange={[Function]}
              placeholder="Search"
              type="text"
              value=""
            />
            <button
              className="m-search-form__clear"
              onClick={[Function]}
              type="reset"
            >
              <span
                className=""
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "SVG",
                  }
                }
              />
            </button>
          </div>
          <button
            className="m-search-form__submit"
            disabled={true}
            onClick={[Function]}
            type="submit"
          >
            <span
              className=""
              dangerouslySetInnerHTML={
                Object {
                  "__html": "SVG",
                }
              }
            />
          </button>
        </div>
        <div
          className="m-search-form__row"
        >
          <ul
            className="m-search-form__property-buttons"
          >
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={true}
                className="m-search-form__property-input"
                id="property-all"
                name="property"
                onChange={[Function]}
                type="radio"
                value="all"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-all"
              >
                all
              </label>
            </li>
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={false}
                className="m-search-form__property-input"
                id="property-run"
                name="property"
                onChange={[Function]}
                type="radio"
                value="run"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-run"
              >
                run
              </label>
            </li>
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={false}
                className="m-search-form__property-input"
                id="property-vehicle"
                name="property"
                onChange={[Function]}
                type="radio"
                value="vehicle"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-vehicle"
              >
                vehicle
              </label>
            </li>
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={false}
                className="m-search-form__property-input"
                id="property-operator"
                name="property"
                onChange={[Function]}
                type="radio"
                value="operator"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-operator"
              >
                operator
              </label>
            </li>
          </ul>
        </div>
      </form>
      <button
        className="m-search-page__toggle-mobile-display-button"
        onClick={[Function]}
      >
        Show 
        map
         instead
      </button>
    </div>
    <div
      className="m-search-display"
    >
      <div
        className="m-recent-searches"
      >
        <div
          className="m-recent-searches__heading"
        >
          Recent Searches
        </div>
      </div>
    </div>
  </div>
  <div
    className="m-search-page__map"
  >
    <div
      className="m-vehicle-map-state m-vehicle-map-state--auto-centering"
    />
    <div
      className="m-vehicle-map"
      id="id-vehicle-map"
    />
  </div>
</div>
`;

exports[`SearchPage renders vehicle data 1`] = `
<div
  className="c-page m-search-page m-search-page--show-list"
>
  <div
    className="m-search-page__input-and-results"
  >
    <div
      className="m-search-page__input"
    >
      <form
        className="m-search-form"
        onSubmit={[Function]}
      >
        <div
          className="m-search-form__row"
        >
          <div
            className="m-search-form__text"
          >
            <input
              autoFocus={true}
              className="m-search-form__input"
              onChange={[Function]}
              placeholder="Search"
              type="text"
              value=""
            />
            <button
              className="m-search-form__clear"
              onClick={[Function]}
              type="reset"
            >
              <span
                className=""
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "SVG",
                  }
                }
              />
            </button>
          </div>
          <button
            className="m-search-form__submit"
            disabled={true}
            onClick={[Function]}
            type="submit"
          >
            <span
              className=""
              dangerouslySetInnerHTML={
                Object {
                  "__html": "SVG",
                }
              }
            />
          </button>
        </div>
        <div
          className="m-search-form__row"
        >
          <ul
            className="m-search-form__property-buttons"
          >
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={true}
                className="m-search-form__property-input"
                id="property-all"
                name="property"
                onChange={[Function]}
                type="radio"
                value="all"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-all"
              >
                all
              </label>
            </li>
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={false}
                className="m-search-form__property-input"
                id="property-run"
                name="property"
                onChange={[Function]}
                type="radio"
                value="run"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-run"
              >
                run
              </label>
            </li>
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={false}
                className="m-search-form__property-input"
                id="property-vehicle"
                name="property"
                onChange={[Function]}
                type="radio"
                value="vehicle"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-vehicle"
              >
                vehicle
              </label>
            </li>
            <li
              className="m-search-form__property-button"
            >
              <input
                checked={false}
                className="m-search-form__property-input"
                id="property-operator"
                name="property"
                onChange={[Function]}
                type="radio"
                value="operator"
              />
              <label
                className="m-search-form__property-label"
                htmlFor="property-operator"
              >
                operator
              </label>
            </li>
          </ul>
        </div>
      </form>
      <button
        className="m-search-page__toggle-mobile-display-button"
        onClick={[Function]}
      >
        Show 
        map
         instead
      </button>
    </div>
    <div
      className="m-search-display"
    >
      <div
        className="m-recent-searches"
      >
        <div
          className="m-recent-searches__heading"
        >
          Recent Searches
        </div>
      </div>
    </div>
  </div>
  <div
    className="m-search-page__map"
  >
    <div
      className="m-vehicle-map-state m-vehicle-map-state--auto-centering"
    />
    <div
      className="m-vehicle-map"
      id="id-vehicle-map"
    />
  </div>
</div>
`;
